torch/csrc/autograd/python_engine.cpp

```diff
@@ -3,13 +3,33 @@
 #include "torch/csrc/autograd/engine.h"
 #include "torch/csrc/THP.h"
 #include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/utils/auto_gil.h"
 
 using namespace torch::autograd;
 
 struct THPEngine {
     PyObject_HEAD
 };
 
+struct PythonEngine : public Engine {
+  virtual void thread_main(ReadyQueue& queue) override {
+    AutoGIL gil;
+    AutoNoGIL no_gil;
```

Comment:
Yes. From the Python source code it looked like the first and last calls to `PyGILState_Ensure` and  `PyGILState_Release` are more expensive since they create and delete `PyThreadState` objects. Without this, internal calls will create and delete `PyThreadState` objects instead of just acquiring and releasing the lock.  It would still be correct without these calls, but may be more expensive.

I'm not entirely sure about the cost -- I haven't measured it without these lines.